src/main/java/io/confluent/connect/elasticsearch/ElasticsearchSinkTask.java

```diff
@@ -15,152 +15,191 @@
-      log.debug("Flushing data to Elasticsearch with the following offsets: {}", offsets);
-      writer.flush();
-    } else {
-      log.debug("Could not flush data to Elasticsearch because ESWriter already closed.");
+  public String version() {
+    return Version.getVersion();
+  }
+
+  private void checkIndex(String index) {
+    if (!indexCache.contains(index)) {
+      log.info("Creating index {}.", index);
+      client.createIndex(index);
+      indexCache.add(index);
     }
   }
 
-  @Override
-  public void close(Collection<TopicPartition> partitions) {
-    log.debug("Closing the task for topic partitions: {}", partitions);
+  private void checkMapping(SinkRecord record) {
+    String index = convertTopicToIndexName(record.topic());
+    if (!config.shouldIgnoreSchema(record.topic()) && !existingMappings.contains(index)) {
+      if (!client.hasMapping(index)) {
+        client.createMapping(index, record.valueSchema());
+      }
+      log.debug("Locally caching mapping for index '{}'", index);
+      existingMappings.add(index);
+    }
   }
 
-  @Override
-  public void stop() throws ConnectException {
-    log.info("Stopping ElasticsearchSinkTask");
-    if (writer != null) {
-      writer.stop();
-      writer = null;
+  /**
+   * Returns the converted index name from a given topic name. Elasticsearch accepts:
+   * <ul>
+   *   <li>all lowercase</li>
+   *   <li>less than 256 bytes</li>
+   *   <li>does not start with - or _</li>
+   *   <li>is not . or ..</li>
+   * </ul>
+   * (<a href="https://github.com/elastic/elasticsearch/issues/29420">ref</a>_.)
+   */
+  private String convertTopicToIndexName(String topic) {
+    String index = topic.toLowerCase();
+    if (index.length() > 255) {
+      index = index.substring(0, 255);
     }
-    if (client != null) {
-      client.close();
+
+    if (index.startsWith("-") || index.startsWith("_")) {
+      index = index.substring(1);
     }
+
+    if (index.equals(".") || index.equals("..")) {
+      index = index.replace(".", "dot");
+      log.warn("Elasticsearch cannot have indices named {}. Index will be named {}.", topic, index);
+    }
+
+    log.trace("Topic '{}' was translated to index '{}'.", topic, index);
+    return index;
+  }
+
+  private void reportBadRecord(SinkRecord record, Throwable error) {
+    if (reporter != null) {
+      reporter.report(record, error);
+    }
+  }
+
+  private boolean shouldSkipRecord(SinkRecord record) {
+    return record.value() == null && config.behaviorOnNullValues() == BehaviorOnNullValues.IGNORE;
+  }
+
+  private void tryWriteRecord(SinkRecord sinkRecord) {
+    DocWriteRequest<?> record = null;
+    try {
+      record = converter.convertRecord(sinkRecord, convertTopicToIndexName(sinkRecord.topic()));
+    } catch (DataException convertException) {
+      reportBadRecord(sinkRecord, convertException);
+
+      if (config.dropInvalidMessage()) {
+        log.error("Can't convert {}.", recordString(sinkRecord), convertException);
```

Comment:
Missing a placeholder?

Reply:
logger can take an exception as the last argument without a placeholder